pandas/core/generic.py

```diff
@@ -6275,11 +6275,13 @@ def fillna(
         else:
             if self.ndim == 1:
                 if isinstance(value, (dict, ABCSeries)):
-                    value = create_series_with_explicit_dtype(
+                    value_map = create_series_with_explicit_dtype(
                         value, dtype_if_empty=object
                     )
-                    value = value.reindex(self.index, copy=False)
-                    value = value._values
+                    value = self.copy()
+                    modification_index = value.index.intersection(value_map.index)
+                    if not modification_index.empty:
+                        value.loc[modification_index] = value_map[modification_index]
```

Comment:
This sounds like a clearer approach than #40509, but still worry about the same comment https://github.com/pandas-dev/pandas/pull/40509/files#r606339592 as its essentially accomplishing the same thing in a different way.

I wonder if the `create_series_with_explicit_dtype` call could be avoided - what you're doing here feels like a `map`, but I guess would need to some tweaking to use here.

Reply:
Also, does this take precedence over #42859? If so, would you mind closing that?

Reply:
I agree, it's very similar to your earlier approach, which is why I tried my other approach first.  This does avoid using any isna check which I was hoping would make it a little more simple/straight-forward and less fragile.

We could also use the value dictionary directly instead of creating a series with create_series_with_explicit_dtype.  This could work by iterating through the value dictionary keys (indexes to be modified) one by one:

```
value_map = value
value = self.copy()
modification_index = value.index.intersection(value_map.keys())
if not modification_index.empty:
    for i in modification_index:
        value.loc[i] = value_map[i]        
```

Reply:
Hmm your original approach seems better than this bc the iteration over index with a `loc` at each step is probably expensive.

Do we have ASV coverage that hits this path? Would want to make sure the current solution doesn't overly deteriorate performance with the added copy and intersection. If it does cause a slowdown, might be worth special-casing object dtype and only taking this more expensive route in that case since multiple nulls are possible

Reply:
@mzeitlin11 Was wondering if you could take a look and let me know what you think?  Also, the CI/ Checks is failing when "Running benchmarks" with "Error: Process completed with exit code 1."  I was wondering if you saw this before or have any ideas on what could be wrong.  I only touched bencharms for fillna and tested it on my local, not sure what would cause the program to exit for what looks like other functions.

Reply:
Taking a look now - looks like failure is for `float32` dtype - does this run fine for you locally? If so, initial guess would be something like a `numpy` version difference causing an issue, will check out locally and see if I can reproduce

Reply:
Seems like a bug - on master
```
ser = pd.Series([1.0, np.nan], dtype="float32")
val = {1: 1.01}
ser.fillna(val)
```
raises 

```
TypeError: Cannot cast array data from dtype('float64') to dtype('float32') according to the rule 'safe'
```

Reply:
Will open an issue for this - for this pr the best approach will probably be to just skip that case in the benchmark. Other examples will have the approach for this, but basically you can raise `NotImplementError` (I think) and it will skip the benchmark

Reply:
@mzeitlin11 I found a way to clean it up a bit by explicitly casting the dtype to avoid this error.  Worked locally, will confirm this works once the CI/CD is complete.